pkg/operator/controller/certificate-publisher/controller.go

```diff
@@ -0,0 +1,160 @@
+// certificates to the "router-certs" secret in the "openshift-config-managed"
+// namespace.
+package certificatepublisher
+
+import (
+	"context"
+	"fmt"
+
+	logf "github.com/openshift/cluster-ingress-operator/pkg/log"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	"k8s.io/client-go/tools/record"
+
+	corev1 "k8s.io/api/core/v1"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	runtimecontroller "sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "certificate-publisher-controller"
+)
+
+var log = logf.Logger.WithName(controllerName)
+
+type reconciler struct {
+	client            client.Client
+	operatorCache     cache.Cache
+	operandCache      cache.Cache
+	recorder          record.EventRecorder
+	operatorNamespace string
+	operandNamespace  string
+}
+
+// New returns a new controller that publishes a "router-certs" secret in the
+// openshift-config-managed namespace with all in-use default certificates.
+func New(mgr manager.Manager, operandCache cache.Cache, cl client.Client, operatorNamespace, operandNamespace string) (runtimecontroller.Controller, error) {
+	operatorCache := mgr.GetCache()
+	reconciler := &reconciler{
+		client:            cl,
+		operatorCache:     operatorCache,
+		operandCache:      operandCache,
+		recorder:          mgr.GetRecorder(controllerName),
+		operatorNamespace: operatorNamespace,
+		operandNamespace:  operandNamespace,
+	}
+	c, err := runtimecontroller.New(controllerName, mgr, runtimecontroller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Index ingresscontrollers over the default certificate name so that
+	// secretIsInUse can look up whether a secret is referenced by any
+	// ingresscontrollers.
+	if err := operatorCache.IndexField(&operatorv1.IngressController{}, "defaultCertificateName", func(o runtime.Object) []string {
+		secret := controller.RouterEffectiveDefaultCertificateSecretName(o.(*operatorv1.IngressController), operandNamespace)
+		return []string{secret.Name}
+	}); err != nil {
+		return nil, fmt.Errorf("failed to create index for ingresscontroller: %v", err)
+	}
+
+	secretsInformer, err := operandCache.GetInformer(&corev1.Secret{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create informer for secrets: %v", err)
+	}
+	if err := c.Watch(&source.Informer{Informer: secretsInformer}, &handler.EnqueueRequestForObject{}, predicate.Funcs{
```

Comment:
It is not _entirely_ obvious why we need `source.Informer` here but can use `source.Kind` below, and the controller-runtime documentation does not explain the difference between the two.  The problem (as best I understand from guessing and then reading the controller-runtime source) is that the `source.Informer` uses the controller's cache, which is for the `openshift-ingress-operator` namespace, and here we need to watch a resource in the `openshift-ingress` namespace.  Does this warrant a comment here? (I probably should submit a docs PR to controller-runtime.)